File: src/main/java/path_generator/discrete_function/DiscreteFunctionBuildings.java
package path_generator.discrete_function;

import jakarta.validation.constraints.NotNull;

import java.util.ArrayList;
import java.util.List;

public class DiscreteFunctionBuildings implements DiscreteFunction<Integer> {
    private final List<Integer> function;
    public static final int FUNCTION_MINIMUM_VALUE = 1;
    public static final int FUNCTION_MAXIMUM_VALUE = 100;
    /**
     * Задает любую функцию с 24 отрезками.
     * @param function
     */
    public DiscreteFunctionBuildings(final @NotNull List<Integer> function) {
        if (function.size() != NUMBER_OF_SECTIONS) {
            throw new IllegalArgumentException("List have size isn't 24.");
        }
        for (int i = 0; i < NUMBER_OF_SECTIONS; i++) {
            Integer currentValue = function.get(i);
            if (currentValue < FUNCTION_MINIMUM_VALUE) {
                throw new IllegalArgumentException("Function value is less than the minimum value.");
            }
            if (currentValue > FUNCTION_MAXIMUM_VALUE) {
                throw new IllegalArgumentException("Function value is greater than the maximum value.");
            }
        }
        this.function = function;
    }

    /**
     * Задает функцию с константным значением.
     * @param constanta
     */
    public DiscreteFunctionBuildings(final int constanta) {
        if (constanta < FUNCTION_MINIMUM_VALUE) {
            throw new IllegalArgumentException("Function value is less than the minimum value.");
        }
        if (constanta > FUNCTION_MAXIMUM_VALUE) {
            throw new IllegalArgumentException("Function value is greater than the maximum value.");
        }
        this.function = new ArrayList<>();
        for (int i = 0; i < NUMBER_OF_SECTIONS; i++) {
            this.function.add(constanta);
        }
    }

    @Override
    public Integer getIthPriority (final int ithHour){
        return function.get(ithHour);
    }
}

File: simulator/src/main/kotlin/path_generator/discrete_function/DiscreteFunctionBuildings.kt
package path_generator.discrete_function

import jakarta.validation.constraints.NotNull

class DiscreteFunctionBuildings : DiscreteFunction<Int?> {
    private val function: MutableList<Int>

    /**
     * Задает любую функцию с 24 отрезками.
     * @param function
     */
    constructor(function: @NotNull MutableList<Int>) {
        require(function.size == DiscreteFunction.NUMBER_OF_SECTIONS) { "List have size isn't 24." }
        for (i in 0 until DiscreteFunction.NUMBER_OF_SECTIONS) {
            val currentValue = function[i]
            require(currentValue >= FUNCTION_MINIMUM_VALUE) { "Function value is less than the minimum value." }
            require(currentValue <= FUNCTION_MAXIMUM_VALUE) { "Function value is greater than the maximum value." }
        }
        this.function = function
    }

    /**
     * Задает функцию с константным значением.
     * @param constanta
     */
    constructor(constanta: Int) {
        require(constanta >= FUNCTION_MINIMUM_VALUE) { "Function value is less than the minimum value." }
        require(constanta <= FUNCTION_MAXIMUM_VALUE) { "Function value is greater than the maximum value." }
        this.function = ArrayList()
        for (i in 0 until DiscreteFunction.NUMBER_OF_SECTIONS) {
            function.add(constanta)
        }
    }

    override fun getIthPriority(ithHour: Int): Int {
        return function[ithHour]
    }

    companion object {
        const val FUNCTION_MINIMUM_VALUE: Int = 1
        const val FUNCTION_MAXIMUM_VALUE: Int = 100
    }
}
